pkg/apis/kudo/v1beta1/instance_admission.go

```diff
@@ -0,0 +1,239 @@
+| HasScheduledPlan | ParameterUpdate | Upgrade | PlanOverride | PlanCancel | Allow |                                                     Description                                                     |
+|------------------|-----------------|---------|--------------|------------|-------|---------------------------------------------------------------------------------------------------------------------|
+| x                | x               |         |              |            | [No]  | Forbid parameter updates when a plan is scheduled unless the same plan is triggered (instance status will be reset) |
+| x                |                 | x       |              |            | No    | Forbid upgrades is another plan is running                                                                          |
+| x                |                 |         | x            |            | No    | Forbid plan overrides (for now)                                                                                     |
+| x                |                 |         |              | x          | No    | Forbid plan cancellations (for now)                                                                                 |
+| ---              |                 |         |              |            |       | ---                                                                                                                 |
+|                  | x               |         | x            |            | No    | Forbid simultaneous parameter update and directly triggered plan                                                    |
+|                  |                 | x       | x            |            | No    | Forbid simultaneous upgrades and directly triggered plans                                                           |
+|                  | x               | x       |              |            | No*   | Forbid simultaneous upgrades and parameter updates                                                                  |
+ --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+ *Note: simultaneous upgrade and parameter update is NOT allowed. However, there is a exception where new OV removes an existing
+ parameter. Removing this parameter in the instance update would not count as parameter update (since there is no plan to trigger).
+
+ For the complete set of rules, see the corresponding test.
+*/
+
+// validateUpdate takes in the old and new (updated) instance and returns a new plan that might
+// be triggered based on the update and an error if the update is not valid. Return plan might be
+// - <nil> when there is no change to an existing scheduled plan
+// - '' empty string when an existing plan should be canceled (not implemented yet)
```

Comment:
uh, this is strange. what maybe we should return a struct then? Having empty string and nil with such a different meanings feel weird

Reply:
hu? why is it strange?
```
"deploy" -> "backup" => override
"deploy" -> "" => cancel
"deploy" -> nil => // no change ¯\_(ツ)_/¯
```
a struct here seems like an overkill.

Reply:
so enum then ? :)

Reply:
I don't know, I always see *string as Option in scala and here you're adding some special meaning to empty string, which feels a bit confusing to me. Normally empty string could mean cancel as well, hard to say 🤷‍♀ I personally would go with enum or something more clear but ... I can let this one pass, even though I think it would lead to nicer API :)